rtl/eclrtl/rtlnewkey.cpp

```diff
@@ -0,0 +1,2133 @@
+    int compareRaw(const RtlTypeInfo & type, const byte * field) const
+    {
+        if (!value)
+        {
+            if (mask & CMPmin)
+                return +1;
+            if (mask & CMPmax)
+                return -1;
+        }
+        return type.compare(field, value.get());
+    }
+
+    int compare(const RtlTypeInfo & type, const byte * field) const
+    {
+        int c = compareRaw(type, field);
+        if (c == 0)
+        {
+            if (mask & CMPeq)
+                return 0;
+            //Lower bound 
+            if (mask & CMPgt)
+                return -1;
+            //Check against upper bound
+            if (mask & CMPlt)
+                return +1;
+            throwUnexpected();
+        }
+        return c;
+    }
+
+    int compareRaw(const RtlTypeInfo & type, const ValueTransition & other) const
+    {
+        if (!value || !other.value)
+        {
+            if (isMinimum())
+                return other.isMinimum() ? 0 : -1;
+
+            if (isMaximum())
+                return other.isMaximum() ? 0 : +1;
+
+            if (other.isMinimum())
+                return +1;
+
+            if (other.isMaximum())
+                return -1;
+
+            throwUnexpected();
+        }
+
+        return type.compare(value, other.value);
+    }
+
+
+    StringBuffer & describe(const RtlTypeInfo & type, StringBuffer & out) const
+    {
+        if (mask & CMPgt)
+        {
+            if (mask & CMPeq)
+                out.append("[");
+            else
+                out.append("(");
+        }
+
+        if (value)
+        {
+            size32_t len;
+            rtlDataAttr text;
+            type.getUtf8(len, text.refstr(), value);
+            size32_t size = rtlUtf8Size(len, text.getstr());
+            if (type.isNumeric())
+            {
+                out.append(size, text.getstr());
+            }
+            else
+            {
+                out.append("'");
+                appendUtf8AsECL(out, size, text.getstr());
+                out.append("'");
+            }
+        }
+
+        if (mask & CMPlt)
+        {
+            if (mask & CMPeq)
+                out.append("]");
+            else
+                out.append(")");
+        }
+        return out;
+    }
+
+    bool isPreviousBound(const RtlTypeInfo & type, const ValueTransition & other) const
+    {
+        if (isInclusiveBound() || other.isInclusiveBound())
+        {
+            if (compareRaw(type, other) == 0)
+                return true;
+        }
+        if (isInclusiveBound() && other.isInclusiveBound())
+        {
+            //MORE: if isPreviousValue(other)
+            //if (oneless(hival, t.getValue()))
+        }
+        return false;
+    }
+
+    //If this transition isn't shared then modify it directly, otherwise clone.  Always return a linked object.
+    ValueTransition * modifyTransition(const RtlTypeInfo & type, TransitionMask newMask)
+    {
+        assertex(value);
+        if (IsShared())
```

Comment:
Any threading concerns doing this?

Reply:
No - access to these sets needs to be single threaded

Reply:
At create time, I can accept (and this is create time I think). I hope they can be accessed from multiple threads after created though!